src/goose/toolkit/vector.py

```diff
@@ -0,0 +1,150 @@
+        file_paths, embeddings = self.load_vector_database(temp_db_path)
+        self.notifier.status("Performing query...")
+        similar_files = self.find_similar_files(query, file_paths, embeddings)        
+        return '\n'.join(similar_files)
+
+    def lookup_db_path(self, repo_path: str) -> str:
+        """
+        Check if a vector database exists for the given repository path or its parent directories.
+
+        Args:
+            repo_path (str): Path to the source code repository.
+
+        Returns:
+            str: Path to the existing vector database file, or None if none found.
+        """
+        current_path = Path(repo_path).expanduser()
+        while current_path != current_path.parent:
+            temp_db_path = self.get_db_path(str(current_path))
+            if os.path.exists(temp_db_path):
+                return temp_db_path
+            current_path = current_path.parent
+        return None
+
+    def scan_repository(self, repo_path):
+        repo_path = Path(repo_path).expanduser()
+        file_contents = []
+        file_paths = []
+        skipped_file_types = {}
+        for root, dirs, files in os.walk(repo_path):
+            # Exclude dotfile directories
+            dirs[:] = [d for d in dirs if not d.startswith('.')]
+            for file in files:
+                file_extension = os.path.splitext(file)[1]
+                if file_extension in ['.py', '.java', '.js', '.jsx', '.ts', '.tsx', '.cpp', '.c', '.h', '.hpp', '.rb', '.go', '.rs', '.php', '.md', '.dart', '.kt', '.swift', '.scala', '.lua', '.pl', '.r', '.m', '.mm', '.f', '.jl', '.cs', '.vb', '.pas', '.groovy', '.hs', '.elm', '.erl', '.clj', '.lisp']:
+                    file_path = os.path.join(root, file)
+                    file_paths.append(file_path)
+                    try:
+                        with open(file_path, 'r', errors='ignore') as f:
+                            content = f.read()
+                            file_contents.append(content)
+                    except Exception as e:
+                        print(f'Error reading {file_path}: {e}')
+                else:
+                    skipped_file_types[file_extension] = True
+        return file_paths, file_contents
+    
+    def build_vector_database(self, file_contents):
+        embeddings = self.model.encode(file_contents, convert_to_tensor=True)
+        return embeddings
+
+    def save_vector_database(self, file_paths, embeddings, db_path):
+        torch.save({'file_paths': file_paths, 'embeddings': embeddings}, db_path)
+
+    def load_vector_database(self, db_path):
+        if db_path is not None and os.path.exists(db_path):
+            data = torch.load(db_path, weights_only=True)
+        else:
+            raise ValueError(f"Database path {db_path} does not exist.")
+        return data['file_paths'], data['embeddings']
+
+    def find_similar_files(self, query, file_paths, embeddings):
```

Comment:
Probably worth adding a dep like `faiss` here to use approximate nearest neighbors?

Reply:
yeah - I wasn't that fussed as it really is just (later on) expanded to paths to look in (with other tools) vs actual direct results (think very large monorepo where you aren't sure where to look to answer general questions)